### DISCUSSION TITLE
Timestamps in the Standard Gleam Time Package

### DISCUSSION THREAD

**Discussion by jrstrunk** [gleam-lang/gleam#4095] [Ideas & suggestions]
Hello all! 

A standard gleam time package is in the works, and I would like to invite community involvement in its development. Its immediate goal is to provide interop between time packages in the ecosystem.

As a first step, the Timestamp type is being explored. Timestamps are just `Int`s, but by giving it a standard type we can help assure that a second denominated timestamp is not accidentally compared to a millisecond denominated timestamp.

The current design is this:
```gleam
pub type Timestamp {
  Timestamp(milliseconds: Int)
}
```

For now, we would like to focus on providing types in the standard package with minimal functionality, relying on the ecosystem to deliver richer functionality. If just core types are provided, then less can be committed to in the standard package. Gleam's type system is unique (and fantastic), and as such we may want to take some novel approaches to dealing with time. The less we commit to the standard library, the more experimentation we can do as a community, and the more cool things we might discover!

Do any of you amazing community members have any additional points we should take into consideration before canonicalizing this design?

After timestamps are settled, we could move on to probably the next simplest type which really isn't all that simple: dates. Some big questions are: 1. how should they be represented? 2. should the constructor be opaque?

I am the author of `tempo` (gtempo on hex), but I would like to make sure @massivefermion (author of `birl`), @maxdeviant (author of `bigben`), and @michaeljones (author of `rada`) are aware of this thread. I am not part of the Gleam core team, so while I have had prior conversations with Louis, nothing I say should be taken as official.

**Comment by jrstrunk** [comment]
These two functions are also being considered, but we may want to keep functionality light so we don't over commit to anything. External packages can also provide additional functions.
```gleam
pub fn to_seconds(timestamp: Timestamp) -> Int {
  timestamp.milliseconds / 1000
}

// Maybe, for use in pipelines mostly I assume
pub fn to_milliseconds(timestamp: Timestamp) -> Int {
  timestamp.milliseconds
}
```

**Comment by massivefermion** [comment]
I think it's better to have an approach that is flexible while at the same time provides the safeguards around units which I think phantom types provide:

```gleam
pub type Unit {
  MicroSecond
  MilliSecond
  Second
  Minute
  Hour
  Day
  Week
  Month
  Year
}

pub type Timestamp(unit) {
  Timestamp(Int)
}
```

We may even omit the `Unit` type itself in the stdlib!

**Reply by maxdeviant** [reply]
Is all that really necessary, though?

For instance, I don't really see a need for a timestamp represented as a certain number of years.

**Reply by massivefermion** [reply]
I just copied the unit type I had. As I said, we may even omit the unit type itself from the stdlib. Although I use that type for my duration module, so larger units are still necessary.

**Reply by giacomocavalieri** [reply]
I don't think having the unit phantom type would make for a particularly good API. For the phantom type to make sense the type would have to be opaque, otherwise people could still do this:
```gleam
let seconds: Timestamp(Second) = Timestamp(1)
let hours: Timestamp(Hour) = Timestamp(2)
seconds.value + hours.value // ???
```
And making it opaque would mean that stdlib would now have to define all the conversion functions to work with those:
```gleam
pub fn seconds_value(timestamp: Timestamp(Second)) -> Int { todo }
pub fn hours_value(timestamp: Timestamp(Hour)) -> Int { todo }
// ... 
```

**Reply by jrstrunk** [reply]
The use of phantom types are cool! But I do have to agree with @giacomocavalieri that they might not fit in with the goals we are going for. Though the point of flexibility is one we should we should make sure we account for.

Lower precision units like `year` are representable in a reasonable way with the current proposed design, but high precision units like `microseconds` and `nanoseconds` are not. Milliseconds were chosen as the unit because they are a common and reasonable default.  Even though they could represent more things, I do not think `microseconds` or `nanoseconds` are appropriate for the default timestamp unit because you start getting into unsafe int ranges on JavaScript when trying to represent the current unix time.

If as a community we later realize we would really like to have a more precise timestamp type, we could introduce a second type that could look something like this (or totally different, but fill the same role):
```gleam
pub type PreciseTimestamp {
  PreciseTimestamp(nanoseconds: Int)
}
```
It seems to me that it would most likely have a totally different use case than the regular 'Timstamp' type, so type fragmentation would probably not be an issue.

Does this path satisfy any long term flexibility concerns?

**Reply by massivefermion** [reply]
I don't think using micro-seconds is problematic. `birl`'s `Time` type actually keeps the timestamp in micro-seconds and that's the unit all the calculations are done in.

**Reply by jrstrunk** [reply]
Ah you're right, the current unix time in microseconds is one decimal away from being unsafe on JS. 

The unit we choose as the underlying will be the easiest to work with, because to get it you just have to call `my_timstamp.milliseconds`, but to get to any other value you'd have to do a conversion like ` my_timstamp.milliseconds / 1000`.

Anecdotally I have seen millisecond timestamps way more often than microsecond timestamps, how should we weigh the cost of convenience for flexibility? Is making the most common use case more cumbersome worth some extra flexibility when there might be other paths to acheive that flexibility? If microseconds was chosen, could we run into issues in the future with the current unix time being so close to unsafe ints in JS?

**Reply by massivefermion** [reply]
I'm not sure about unsafe ints in js. As far as I know, it's not affecting anything `birl` does at all. But I didn't suggest it only because of flexibility. The main reason I used micro-seconds in `birl` was consistency and accuracy. When numbers get large, you can't really trust the less significant digits. So by using micro-seconds for calculations but mostly using milli-seconds for the interface, you're making sure any inconsistency or inaccuracy is at a level that will most likely stay hidden to the user.

**Reply by jrstrunk** [reply]
What about using a float for the milliseconds value? Floats in erlang and Numbers in JS are both IEEE 754 64-bit floating point numbers, so if the type's only property was `milliseconds: Float` then we could keep the convenience of having milliseconds easy to work with while allowing for more precision if needed. I know floats can be tricky though, and I don't think I can definitely say this is a good or bad idea.

JavaScript's [Performance.now()](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) returns a millisecond value as a float, so seems like someone over there determined it was a good idea for some reason.

Would love to hear some more opinions on this.

**Reply by massivefermion** [reply]
I don't think there is any issue here. The time for which the timestamp in microseconds would exceed the safe int range for js is hundreds of years from now. And even for times after that, it will automatically switch to floating points which means there will be some inaccuracies, but again those inaccuracies are in the micro-seconds range which won't affect user facing values if all of those are presented in milli-seconds. I'm pretty sure we can safely use micro-seconds for underlying calculations.

**Reply by jrstrunk** [reply]
Those are good points. Also, PostgreSQL stores timestamps in microseconds, and I assume storing time in a database is going to be one of the large use cases of timestamps. 

I also want to consider how this fits in with Gleam, not just programming in general. Gleam seems to lean towards correctness and explicitness over convenience and conciseness; I think using microseconds as the underlying value fits in fine with those values.

Another point is that timestamps are not human-readable by nature, moving from milliseconds to microseconds does not change that. No matter which unit is chosen, users are going to need functions to make timestamps more presentable.

I am fine with canonicalizing `microseconds: Int`!

**Reply by lpil** [reply]
> The time for which the timestamp in microseconds would exceed the safe int range for js is hundreds of years from now.

That doesn't seem very far away. I have used dates many years in the future before.

**Reply by jrstrunk** [reply]
It is 230 years away, that is less than what I would have guessed. I agree it is too close for comfort, but I also see a need for representing time in microseconds in the standard package.

What if the type looked like this:
```gleam
pub opaque type Timestamp {
  // Might as well go to nanosecond precision at this point
  Timestamp(day: Int, nanoseconds: Int)
}
```
Where the day `Int` value represents the number of days since Jan 1st, 1970 and the nanoseconds are only the nanoseconds since midnight on that day (always safe in JS). This would provide a standard type for nearly all use cases at the cost of making the type opaque (less friendly). (I am not sure if there is a real broad performance decrease because this makes sub-day arithmetic and converting from Int timestamps slower while making day arithmetic and formatting faster).

We would also then want to make nice constructing functions for the type like `timestamp.from_milliseconds`.

**Reply by maxdeviant** [reply]
Apropos overflowing nanoseconds, here's a real-world example in queuing backoff: https://brandur.org/fragments/go-max-time-duration

**Reply by massivefermion** [reply]
> > The time for which the timestamp in microseconds would exceed the safe int range for js is hundreds of years from now.
> 
> That doesn't seem very far away. I have used dates many years in the future before.

Does `birl` have issues for those workflows?

**Reply by lpil** [reply]
@jrstrunk What was your reason for going for days there? Anything in particular? Or just a semi-arbitrary decision and it could have been years or something else?

**Reply by jrstrunk** [reply]
@maxdeviant cool piece of context, good to keep in mind. @massivefermion I believe it does, but you may want to write some tests to verify you lose duration precision past year 2255 something.

@lpil it was semi-arbitrary, we could also do something like this:
```gleam
pub opaque type Timestamp {
  // milliseconds is the number of milliseconds since Jan 1st, 1970
  // nanoseconds is the number of nanoseconds since the last millisecond (never surpasses 100,000)
  Timestamp(milliseconds: Int, nanoseconds: Int)
}
```
This would optimize for millisecond operations (which I assume will be very common) while allowing for nanosecond precision if needed.

**Reply by lpil** [reply]
I suppose we would do a bit of maths to work out the precision limits of a few options and then decide if there's any particular advantage to any of them.

**Reply by jrstrunk** [reply]
A timestamp based on milliseconds should be safe for another 30,000 years, and a timestamp based on seconds should be safe for another 30,000,000 years. The performance difference between something like 
```gleam
pub opaque type Timestamp {
  Timestamp(milliseconds: Int, nanoseconds: Int)
}
```
and something like 
```gleam
pub opaque type Timestamp {
  Timestamp(days: Int, nanoseconds: Int)
}
```
for any given function should not be more than 1 div or mul operation, so less than a nanosecond on modern machines. The only exception to this would be add or subtract operations to the lesser value, leading to an update of the greater value (eg. when `nanoseconds` is added to and becomes larger than 100,000, then the `milliseconds` value has to be updated and the `nanoseconds` value has to be recalculated). 

Also if we make the constructor opaque, couldn't we change around some inner workings later if a compelling reason emerges? (Keeping the same functions of course.)

**Comment by lpil** [comment]
The Elm time library has some other functions that make use of zones. Would we want to adapt those too?

https://package.elm-lang.org/packages/elm/time/latest/Time

If we do not provide anything along these lines how would the programmer be expected to work with these timestamps? Being able to convert them to a textual representation seems like it would be one of the most common needs.

**Reply by massivefermion** [reply]
Not to be pedantic but people usually use timezone and offset interchangeably but we shouldn't be doing that here. If you actually mean timezones, then that should be a long term goal given that we need to find a way to incorporate a timezone db that can be easily updated.
But if you mean offset, then of course, if we're going to convert to and from strings, we should be able to handle offsets from the start.

**Reply by lpil** [reply]
The linked code uses zones, not offsets. I did not bring up offsets so I'm not sure what you mean.

Are you suggested we should not support zones and instead support offsets? Wouldn't it be challenging to determine the appropriate offset?

**Reply by massivefermion** [reply]
Actually no. It's way easier to determine the offset than to determine the timezone and then infer offset from it. `birl` had support for offsets from day one but even now the timezone support is not that satisfactory.

**Reply by lpil** [reply]
Cool! If I have a timestamp how would I know what the offset would need to be to print it appropriately for London?

**Reply by massivefermion** [reply]
ok, I see what you mean now. I was talking about determining the offset of the host system. As I said, determining the timezone itself is just step one which is already solved in `birl`.

The second step, inferring offset from timezone, is actually the hard problem that I've been struggling with. Right now `birl` has only a list of constant offsets for each timezone but offsets are not constant. A country may follow DST which means seasons affect the offset(countries may also opt in or out of DST as they please, so even that is not consistent). There are also other bureaucratic reasons for the offsets to change, so basically what offset corresponds to a timezone depends on what the time is!

So how do people do it? There is a timezone database maintained by IANA. By parsing this database, you can have all the information necessary to do it, but that's not a simple task. Right now all I do with that database is getting the latest offset for each timezone. It takes some research to figure out how to make full use of it.

**Reply by lpil** [reply]
That's right, it's normal to embed the time zone database.

Do you have thoughts on my original question?

**Reply by massivefermion** [reply]
I'm completely in favor of supporting timezones, but at first we need to put some time into learning how to fully utilize the IANA database. Then we need to figure how to keep the timezone database updated.
Right now I run a little bit of rust to download the database, parse it and write the latest offsets in a gleam file as a keyword list, which means every time you want to update the offsets, you need to publish a new release, which does not seem like a sustainable solution to me.

**Reply by giacomocavalieri** [reply]
So we have some data to make an informed decision, this is a table with the number of releases per year the database has gone through. It looks like, on busiest years, it had a new version roughly every month or so

```txt
year releases
2012        7
2013        9
2014       10
2015        7
2016       10
2017        3
2018        7
2019        3
2020        6
2021        5
2022        7
2023        4
2024        2
```

**Reply by jrstrunk** [reply]
Just for some clarification, those Elm functions are taking a integer (assuming it is in UTC), converting it to the given time zone, then returning a different integer (then converting it to a month or whatever else). It is just a transformation on a naive int type. We could implement those Elm functions with the current proposed timestamp type in the original post.

Are you asking if we should add more functions to the package, or asking if we should add more properties to the type?

If we are just discussing adding functions, then I'd quickly say yes eventually those would be nice to add. If we are discussing adding more properties to the type, then there is a lot to talk over.

**Reply by lpil** [reply]
The frequency with which it is updated doesn't really matter as it's unavoidable. The database will need to be packaged for Gleam programs.

@jrstrunk not sure what you mean- they don't take an int, there is a zone type.

I don't have any ideas for additional fields, what did you have in mind?

If we were not to provide any basic functions for working with timestamps then birl or gtempo would presumably need to add these for the timestamp type to be usable at all. Is there a benefit to doing it that way?

**Reply by jrstrunk** [reply]
I guess I meant the function takes an Int (the Elm Posix type) and a time zone and returns an Int. The time zone data isn't tied to the return data type.

I think the benefit of offloading the logic to Birl and Tempo (and maybe more) is that it gives the community much more time to design and experiment with what will eventually become the standard types and functions. I'd like to see the standard be really thought out and done in a Gleamy way :)

**Reply by kwando** [reply]
It would be annoying to have to rebuild your app to get the latest timezone information. Would be cool if it could be dynamic somehow. Maybe look at how it is done in Elixir, which I personally think is quite good.

**Reply by lpil** [reply]
I want to take the same approach as Elixir and Elm and have no restrictions on where zone information comes from, so each application can decide what's best for them.

**Comment by michaeljones** [comment]
I appreciate being included in the initial message but `rada` is a line by line port of `justinmimbs/date` from Elm which I did because I liked the API and it had served me well in my Elm projects. I don't have any particular knowledge to share from this experience. Standardising date and time types seems like a good idea though!

**Comment by lpil** [comment]
I have started on a time library here: https://github.com/gleam-lang/time/pull/1

**Reply by jrstrunk** [reply]
Exciting! I'll put any discussion points as separate issues there so they can be discussed and resolved individually
